src/main/java/org/broadinstitute/hellbender/utils/mcmc/GibbsSampler.java

```diff
@@ -0,0 +1,102 @@
+
+    private final int numSamples;
+    private int numSamplesPerLogEntry;
+
+    private final ParameterizedModel<S, T> model;
+
+    private final List<S> samples;
+
+    private boolean isMCMCRunComplete = false;
+
+    /**
+     * Constructs a GibbsSampler given the total number of samples (including burn-in) and a {@link ParameterizedModel}.
+     * The {@link ParameterizedState} held by the model is used to initialize the Monte Carlo Markov Chain and is taken
+     * to be the first sample.  Number of samples per log entry will be set to the default.
+     * @param numSamples    total number of samples; must be positive
+     * @param model         {@link ParameterizedModel} to be sampled
+     */
+    public GibbsSampler(final int numSamples, final ParameterizedModel<S, T> model) {
+        if (numSamples <= 0) {
+            throw new IllegalArgumentException("Number of samples must be positive.");
+        }
+        this.numSamples = numSamples;
+        this.numSamplesPerLogEntry = NUMBER_OF_SAMPLES_PER_LOG_ENTRY;
+        this.model = model;
+        samples = new ArrayList<>(numSamples);
+        samples.add(model.state());
+    }
+
+    /**
+     * Changes the number of samples per log entry.
+     * @param numSamplesPerLogEntry number of samples per log entry; must be positive
+     */
+    public void setNumSamplesPerLogEntry(final int numSamplesPerLogEntry) {
+        if (numSamplesPerLogEntry <= 0) {
+            throw new IllegalArgumentException("Number of samples per log entry must be positive.");
+        }
+        this.numSamplesPerLogEntry = numSamplesPerLogEntry;
+    }
+
+    /**
+     * Runs the Monte Carlo Markov Chain, using the state of the model provided in the constructor to initialize.
+     * Progress is logged according to {@code numSamplesPerLogEntry}.
+     */
+    public void runMCMC() {
+        logger.info("Starting MCMC sampling.");
+        for (int sample = 1; sample < numSamples; sample++) {
+            if (sample % numSamplesPerLogEntry == 0) {
+                logger.info(sample + " of " + numSamples + " samples generated.");
+            }
+            model.update(rng);
+            samples.add(model.state());
+        }
+        logger.info(numSamples + " of " + numSamples + " samples generated.");
+        logger.info("MCMC sampling complete.");
+        isMCMCRunComplete = true;
+    }
+
+    /**
+     * Returns a list of samples for a specified model parameter, discarding the first {@code numBurnIn} samples.
+     * @param parameterName         name of parameter
+     * @param parameterValueClass   class of parameter value
+     * @param numBurnIn             number of burn-in samples to discard from beginning of chain
+     * @param <U>                   type of parameter value
+     * @return                      List of parameter samples
+     */
+    public <U> List<U> getSamples(final String parameterName, final Class<U> parameterValueClass, final int numBurnIn) {
```

Comment:
tried to be consistent about using S/S1/S2 for AbstractParameterizedState, T/T1/T2 for DataCollection, and U for parameter values...if this is confusing and doesn't add to the readability i can change this to T